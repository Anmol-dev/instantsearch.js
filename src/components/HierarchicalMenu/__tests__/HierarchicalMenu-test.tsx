/** @jsx h */

import { h } from 'preact';
import { render, fireEvent } from '@testing-library/preact';
import HierarchicalMenu from '../HierarchicalMenu';
import itemsFixture from './fixtures/items.json';

const cssClasses = {
  root: 'root',
  noRefinementRoot: 'noRefinementRoot',
  list: 'list',
  childList: 'childList',
  item: 'item',
  selectedItem: 'selectedItem',
  parentItem: 'parentItem',
  link: 'link',
  label: 'label',
  count: 'count',
  showMore: 'showMore',
  disabledShowMore: 'disabledShowMore',
};

const templates = {
  item:
    '<a class="{{cssClasses.link}}" href="{{url}}">' +
    '<span class="{{cssClasses.label}}">{{label}}</span>' +
    '<span class="{{cssClasses.count}}">{{#helpers.formatNumber}}{{count}}{{/helpers.formatNumber}}</span>' +
    '</a>',
  showMoreText: `
    {{#isShowingMore}}
      Show less
    {{/isShowingMore}}
    {{^isShowingMore}}
      Show more
    {{/isShowingMore}}
  `,
};

const defaultProps = {
  canToggleShowMore: false,
  createURL: value => `#${value}`,
  cssClasses,
  hasExhaustiveItems: true,
  isShowingMore: false,
  items: [],
  showMore: false,
  templateProps: {
    templates,
  },
  toggleRefinement: () => {},
  toggleShowMore: () => {},
};

describe('HierarchicalMenu', () => {
  test('renders correct DOM without items', () => {
    const props = {
      ...defaultProps,
    };

    const { container } = render(<HierarchicalMenu {...props} />);
    const list = container.querySelector('.list');

    expect(list).not.toBeInTheDocument();
    expect(container).toMatchSnapshot();
  });

  test('renders correct DOM with items', () => {
    const items = itemsFixture;

    const props = {
      ...defaultProps,
      items,
    };

    const { container } = render(<HierarchicalMenu {...props} />);
    const list = container.querySelector('.list');
    const listItems = container.querySelectorAll('.item');

    expect(list).toBeInTheDocument();
    // There are 20 items because:
    // - 10 level 1
    // - 10 level 2 in the first level 1
    expect(listItems).toHaveLength(20);
    expect(container).toMatchSnapshot();
  });

  test('calls templates with correct data', () => {
    const itemTemplate = jest.fn(() => '');
    const props = {
      ...defaultProps,
      items: [
        {
          label: 'Video Games',
          value: 'Video Games',
          count: 505,
          isRefined: false,
          exhaustive: true,
          data: null,
        },
      ],
      templateProps: {
        templates: {
          ...defaultProps.templateProps.templates,
          item: itemTemplate,
        },
      },
    };

    render(<HierarchicalMenu {...props} />);

    expect(itemTemplate).toHaveBeenCalledTimes(1);
    expect(itemTemplate).toHaveBeenCalledWith({
      count: 505,
      cssClasses,
      data: null,
      exhaustive: true,
      isRefined: false,
      label: 'Video Games',
      url: '#Video Games',
      value: 'Video Games',
    });
  });

  test('triggers refinement on click', () => {
    const toggleRefinement = jest.fn();
    const props = {
      ...defaultProps,
      items: [
        {
          label: 'Video Games',
          value: 'Video Games',
          count: 505,
          isRefined: false,
          exhaustive: true,
          data: null,
        },
      ],
      toggleRefinement,
    };

    const { container } = render(<HierarchicalMenu {...props} />);
    const item = container.querySelector('.item')!;

    fireEvent.click(item);

    expect(toggleRefinement).toHaveBeenCalledTimes(1);
    expect(toggleRefinement).toHaveBeenCalledWith('Video Games');
  });

  test('triggers refinement on a child click', () => {
    const toggleRefinement = jest.fn();
    const props = {
      ...defaultProps,
      items: [
        {
          label: 'Appliances',
          value: 'Appliances',
          count: 4306,
          isRefined: true,
          exhaustive: true,
          data: [
            {
              label: 'Dishwashers',
              value: 'Appliances > Dishwashers',
              count: 181,
              isRefined: false,
              exhaustive: true,
              data: null,
            },
          ],
        },
      ],
      toggleRefinement,
    };

    const { getByText } = render(<HierarchicalMenu {...props} />);
    const childItem = getByText('Dishwashers');

    fireEvent.click(childItem);

    expect(toggleRefinement).toHaveBeenCalledTimes(1);
    expect(toggleRefinement).toHaveBeenCalledWith('Appliances > Dishwashers');
  });

  test('renders with showMore and isShowingMore to false', () => {
    const props = {
      ...defaultProps,
      showMore: true,
      isShowingMore: false,
      canToggleShowMore: true,
    };

    const { container } = render(<HierarchicalMenu {...props} />);
    const showMore = container.querySelector('.showMore');

    expect(showMore).toBeInTheDocument();
    expect(showMore).toHaveTextContent('Show more');
    expect(container).toMatchSnapshot();
  });

  test('renders with showMore and isShowingMore to true', () => {
    const props = {
      ...defaultProps,
      showMore: true,
      isShowingMore: true,
      canToggleShowMore: true,
    };

    const { container } = render(<HierarchicalMenu {...props} />);
    const showMore = container.querySelector('.showMore');

    expect(showMore).toBeInTheDocument();
    expect(showMore).toHaveTextContent('Show less');
    expect(container).toMatchSnapshot();
  });

  test('renders with showMore and canToggleShowMore to false', () => {
    const props = {
      ...defaultProps,
      showMore: true,
      isShowingMore: false,
    };

    const { container } = render(<HierarchicalMenu {...props} />);
    const showMore = container.querySelector('.showMore');

    expect(showMore).toBeInTheDocument();
    expect(showMore).toHaveTextContent('Show more');
    expect(showMore).toBeDisabled();
    expect(container).toMatchSnapshot();
  });

  test('calls toggleShowMore on show more button click', () => {
    const props = {
      ...defaultProps,
      showMore: true,
      isShowingMore: false,
      canToggleShowMore: true,
      toggleShowMore: jest.fn(),
    };

    const { container } = render(<HierarchicalMenu {...props} />);
    const showMore = container.querySelector('.showMore')!;

    fireEvent.click(showMore);

    expect(props.toggleShowMore).toHaveBeenCalledTimes(1);
  });
});
