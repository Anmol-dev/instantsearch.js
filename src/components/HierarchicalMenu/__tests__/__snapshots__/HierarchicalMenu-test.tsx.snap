// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HierarchicalMenu renders correct DOM with items 1`] = `
<div>
  <div
    class="root"
  >
    <ul
      class="list"
    >
      <li
        class="item selectedItem parentItem"
      >
        <div>
          <a
            class="link"
            href="#Appliances"
          >
            <span
              class="label"
            >
              Appliances
            </span>
            <span
              class="count"
            />
          </a>
        </div>
        <div
          class="childList"
        >
          <ul
            class="list"
          >
            <li
              class="item"
            >
              <div>
                <a
                  class="link"
                  href="#Appliances > Dishwashers"
                >
                  <span
                    class="label"
                  >
                    Dishwashers
                  </span>
                  <span
                    class="count"
                  />
                </a>
              </div>
            </li>
            <li
              class="item"
            >
              <div>
                <a
                  class="link"
                  href="#Appliances > Fans"
                >
                  <span
                    class="label"
                  >
                    Fans
                  </span>
                  <span
                    class="count"
                  />
                </a>
              </div>
            </li>
            <li
              class="item"
            >
              <div>
                <a
                  class="link"
                  href="#Appliances > Freezers & Ice Makers"
                >
                  <span
                    class="label"
                  >
                    Freezers & Ice Makers
                  </span>
                  <span
                    class="count"
                  />
                </a>
              </div>
            </li>
            <li
              class="item"
            >
              <div>
                <a
                  class="link"
                  href="#Appliances > Heaters"
                >
                  <span
                    class="label"
                  >
                    Heaters
                  </span>
                  <span
                    class="count"
                  />
                </a>
              </div>
            </li>
            <li
              class="item"
            >
              <div>
                <a
                  class="link"
                  href="#Appliances > Humidifiers"
                >
                  <span
                    class="label"
                  >
                    Humidifiers
                  </span>
                  <span
                    class="count"
                  />
                </a>
              </div>
            </li>
            <li
              class="item"
            >
              <div>
                <a
                  class="link"
                  href="#Appliances > Microwaves"
                >
                  <span
                    class="label"
                  >
                    Microwaves
                  </span>
                  <span
                    class="count"
                  />
                </a>
              </div>
            </li>
            <li
              class="item"
            >
              <div>
                <a
                  class="link"
                  href="#Appliances > Ranges, Cooktops & Ovens"
                >
                  <span
                    class="label"
                  >
                    Ranges, Cooktops & Ovens
                  </span>
                  <span
                    class="count"
                  />
                </a>
              </div>
            </li>
            <li
              class="item"
            >
              <div>
                <a
                  class="link"
                  href="#Appliances > Refrigerators"
                >
                  <span
                    class="label"
                  >
                    Refrigerators
                  </span>
                  <span
                    class="count"
                  />
                </a>
              </div>
            </li>
            <li
              class="item"
            >
              <div>
                <a
                  class="link"
                  href="#Appliances > Small Kitchen Appliances"
                >
                  <span
                    class="label"
                  >
                    Small Kitchen Appliances
                  </span>
                  <span
                    class="count"
                  />
                </a>
              </div>
            </li>
            <li
              class="item"
            >
              <div>
                <a
                  class="link"
                  href="#Appliances > Washers & Dryers"
                >
                  <span
                    class="label"
                  >
                    Washers & Dryers
                  </span>
                  <span
                    class="count"
                  />
                </a>
              </div>
            </li>
          </ul>
        </div>
      </li>
      <li
        class="item"
      >
        <div>
          <a
            class="link"
            href="#Audio"
          >
            <span
              class="label"
            >
              Audio
            </span>
            <span
              class="count"
            />
          </a>
        </div>
      </li>
      <li
        class="item"
      >
        <div>
          <a
            class="link"
            href="#Cameras & Camcorders"
          >
            <span
              class="label"
            >
              Cameras & Camcorders
            </span>
            <span
              class="count"
            />
          </a>
        </div>
      </li>
      <li
        class="item"
      >
        <div>
          <a
            class="link"
            href="#Car Electronics & GPS"
          >
            <span
              class="label"
            >
              Car Electronics & GPS
            </span>
            <span
              class="count"
            />
          </a>
        </div>
      </li>
      <li
        class="item"
      >
        <div>
          <a
            class="link"
            href="#Cell Phones"
          >
            <span
              class="label"
            >
              Cell Phones
            </span>
            <span
              class="count"
            />
          </a>
        </div>
      </li>
      <li
        class="item"
      >
        <div>
          <a
            class="link"
            href="#Computers & Tablets"
          >
            <span
              class="label"
            >
              Computers & Tablets
            </span>
            <span
              class="count"
            />
          </a>
        </div>
      </li>
      <li
        class="item"
      >
        <div>
          <a
            class="link"
            href="#Health, Fitness & Beauty"
          >
            <span
              class="label"
            >
              Health, Fitness & Beauty
            </span>
            <span
              class="count"
            />
          </a>
        </div>
      </li>
      <li
        class="item"
      >
        <div>
          <a
            class="link"
            href="#Office & School Supplies"
          >
            <span
              class="label"
            >
              Office & School Supplies
            </span>
            <span
              class="count"
            />
          </a>
        </div>
      </li>
      <li
        class="item"
      >
        <div>
          <a
            class="link"
            href="#TV & Home Theater"
          >
            <span
              class="label"
            >
              TV & Home Theater
            </span>
            <span
              class="count"
            />
          </a>
        </div>
      </li>
      <li
        class="item"
      >
        <div>
          <a
            class="link"
            href="#Video Games"
          >
            <span
              class="label"
            >
              Video Games
            </span>
            <span
              class="count"
            />
          </a>
        </div>
      </li>
    </ul>
  </div>
</div>
`;

exports[`HierarchicalMenu renders correct DOM without items 1`] = `
<div>
  <div
    class="root noRefinementRoot"
  />
</div>
`;

exports[`HierarchicalMenu renders with showMore and canToggleShowMore to false 1`] = `
<div>
  <div
    class="root noRefinementRoot"
  >
    <button
      class="showMore disabledShowMore"
      disabled=""
    >
      Show more
    </button>
  </div>
</div>
`;

exports[`HierarchicalMenu renders with showMore and isShowingMore to false 1`] = `
<div>
  <div
    class="root noRefinementRoot"
  >
    <button
      class="showMore"
    >
      Show more
    </button>
  </div>
</div>
`;

exports[`HierarchicalMenu renders with showMore and isShowingMore to true 1`] = `
<div>
  <div
    class="root noRefinementRoot"
  >
    <button
      class="showMore"
    >
      Show less
    </button>
  </div>
</div>
`;
